### PR TITLE
Additional crystals added to the EcalBadCal MET filter

### PR DESCRIPTION
**Pull request by lathomas**
New hot EE crystals have been recently observed in 2017 and 2018. 
After discussion between ECAL and JME it was decided to add them to the list of crystal considered by the ecal bad calibration MET filter. 

More information here: 
https://indico.cern.ch/event/775564/contributions/3239371/attachments/1765837/2866952/ecalbadcalfilterupdate.pdf
https://twiki.cern.ch/twiki/bin/viewauth/CMS/MissingETOptionalFiltersRun2#How_to_run_ecal_BadCalibReducedM 

@ahinzmann @peruzzim @eioannou @zdemirag

### CODE REVIEW

**Comment by peruzzim** [comment]
Please ask for advice and agree on how to extend the crystal list also in view of backport to 10_2_X, where you will not be allowed to just extend it like that because it would change the content of miniAOD.

**Comment by perrotta** [comment]
please test

**Comment by lathomas** [comment]
@peruzzim I updated the config file and added a new instance of the filter to be run on MINIAOD. 
From what I understand this can be safely backported to 102X and 94X. Can you confirm this is fine? 
If yes I'll open the PR.

File: RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py [hunk 1]
```diff
@@ -29,4 +29,39 @@
...
+  baddetEcal       = cms.vuint32([872439604,872422825,872420274,872423218,
+                                  872423215,872416066,872435036,872439336,
+                                  872420273,872436907,872420147,872439731,
+                                  872436657,872420397,872439732,872439339,
+                                  872439603,872422436,872439861,872437051,
+                                  872437052,872420649,872422436,872421950,
+                                  872437185,872422564,872421566,872421695,
+                                  872421955,872421567,872437184,872421951,
+                                  872421694,872437056,872437057,872437313]),
+  taggingMode = cms.bool(True),
```
Comment (peruzzim) [inline RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py]: I would suggest to set this to False by default, as it is for the filter above

File: RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py [hunk 2]
```diff
@@ -29,4 +29,39 @@
...
                                                        872436657,872420397,872439732,872439339,
-                                                       872439603,872422436])
+                                                       872439603,872422436,872439861,872437051,
+                                                       872437052,872420649,872422436,872421950,
+                                                       872437185,872422564,872421566,872421695,
+                                                       872421955,872421567,872437184,872421951,
+                                                       872421694,872437056,872437057,872437313])
+
+
+ecalBadCalibReducedMINIAODFilter = cms.EDFilter(
```
Comment (peruzzim) [inline RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py]: I would suggest giving this a more peculiar name, as there is no a-priori reason to define bad crystals in a different way when running on AOD or MINIAOD.
I would name it ecalBadCalibReducedMINIAODPre104XFilter to make it clear it has to be run and considered only when running on MINIAOD produced in older releases.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@peruzzim @fgolf 
this is not for nanoAOD production just yet, right?
Please confirm.
Thank you.

**Comment by peruzzim** [comment]
@slava77 we were just going to comment, yes it would be very nice to have it.
Sorry again for the late notice, it is something that has emerged very recently.

**Comment by slava77** [comment]
> @slava77 we were just going to comment, yes it would be very nice to have it.

please clarify the meaning of this for this PR.
- the ecalBadCalibFilter runs only in miniAOD ==> will not be used in the upcoming 102X nanoAOD
- the new ecalBadCalibReducedMINIAODFilter is not added to any of the tasks/sequences in nanoAOD to create a usable variable.

So, presently this PR is not a part of nanoAOD processing

**Comment by peruzzim** [comment]
- ecalBadCalibFilter will not be used in 102X nanoAOD, but will have to be used from now on in productions to happen in 104X and later, with the updated list of bad xtals. So we are updating that filter to reject more xtals from now on.
- we want to add a filter cutting on the additional xtals in 102X nanoAOD, but we cannot just modify ecalBadCalibFilter because it would change the content of miniAOD in 102X (production). So we add an additional filter that will not be in the 102X miniAOD sequence, and will be called only from nanoAOD (PR will come as soon as we define its name).
- we have to support running on miniAOD produced in 103X and earlier releases, when running from a 104X area. So, in those cases, we want to run the filter on the additional xtals, on the reduced rechit collection. This could in principle be avoided by cloning ecalBadCalibFilter and changing the source collection, would you prefer that rather than having this filter defined in 104X+?

Only the update of ecalBadCalibReducedMINIAODFilter will be backported to 102X.

**Comment by slava77** [comment]
> that will not be in the 102X miniAOD sequence, and will be called only from nanoAOD (PR will come as soon as we define its name).

I prefer this to happen in one PR

**Comment by peruzzim** [comment]
@slava77 sorry can you please detail how you would like us to do in order to support the use cases listed above?
What we would do in nanoAOD is pretty clear, we want to run the updated filter on the miniAOD xtal collection and save a bool with its result.
Can you please suggest your preferred implementation of this? Should we define a catch-all modifier like run2_miniaod_postAutumn18 and start to put all these things in there? I don't think we can define a modifier for every thing we're going to change in 102X in the next months.

File: RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py [hunk 3]
```diff
@@ -29,4 +29,39 @@
...
                                                        872436657,872420397,872439732,872439339,
-                                                       872439603,872422436])
+                                                       872439603,872422436,872439861,872437051,
+                                                       872437052,872420649,872422436,872421950,
+                                                       872437185,872422564,872421566,872421695,
+                                                       872421955,872421567,872437184,872421951,
+                                                       872421694,872437056,872437057,872437313])
+
+
+ecalBadCalibReducedMINIAODPre104XFilter = cms.EDFilter(
```
Comment (slava77) [inline RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py]: The settings should not be copy-pasted from what is in ecalBadCalibFilter.
Presently, ```ecalBadCalibReducedMINIAODPre104XFilter = ecalBadCalibFilter.clone(EcalRecHitSource = 'reducedEgamma:reducedEERecHits')``` should give identical result.

I don't understand the need to have "Pre104X" in the name. IIUC, the simplest is to just apply it to all data anyways on top of the default ecalBadCalibFilter. Either it's an analysis running on miniAOD made with the most up-to-date list of xtals [in this case this filter is redundant or just a subset of the full one ecalBadCalibFilter] or your miniAOD doesn't cover all xtals in which this ecalBadCalibReducedMINIAODFilter triggers (partially/the best way possible)  on the bad events.

Also, nominally, we do not allow multiple modules defined in cfi files.
So, this definition should be moved to a separate _cff file.

**Comment by peruzzim** [comment]
Ok, agreed for 104X, thanks.
How should we do to backport this to 102X, preserving the default miniAOD behaviour there but allowing people (including nanoAOD producer, but also in general all analysers) to call it with the updated xtal list when running on miniAOD?
Thanks a lot for your help!

**Comment by slava77** [comment]
> @slava77 sorry can you please detail how you would like us to do in order to support the use cases listed above?
> What we would do in nanoAOD is pretty clear, we want to run the updated filter on the miniAOD xtal collection and save a bool with its result.
> Can you please suggest your preferred implementation of this? Should we define a catch-all modifier like run2_miniaod_postAutumn18 and start to put all these things in there? I don't think we can define a modifier for every thing we're going to change in 102X in the next months.

As we discussed in the RECO/AT meeting last week, there should be a "future miniAOD" modifier for all things proposed for a backport and modifying the default/available miniAOD results.
I was thinking ```run2_miniAOD_devel``` could be a good name.
This modifier will be empty in the master and will capture changes in not-yet-defined miniAOD in the production release.

**Comment by slava77** [comment]
On 12/6/18 11:21 AM, peruzzim wrote:
> Ok, agreed for 104X, thanks.
> How should we do to backport this to 102X, preserving the default
> miniAOD behaviour there but allowing people (including nanoAOD producer,
> but also in general all analysers) to call it with the updated xtal list
> when running on miniAOD?
> Thanks a lot for your help!
> 
> —

looking at what was done before,
I think that there should be something like run2_nanoAOD_102Xv1 modifier
for 2018 inputs
in addition to what you have for 94X miniAOD (the 2016 "v3" and 2017 "v2").
This shall formalize support of the 102X miniAOD format in the master branch
and can be used explicitly in the 102X backports as well.
Simply having everything in just Run2_2018 era will lead to problems for
simultaneous support in 102X and 104X/master because the miniAOD in the
master will be evolving.

**Comment by peruzzim** [comment]
Thanks, we're doing as you suggest #25443 #25444.
An update of this PR, #25372 and #25368 will follow once the modifiers are merged.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by peruzzim** [comment]
@lathomas this PR needs to be also backported to 102X, modifying (only in the backport, the master looks good as it is) the list of xtals only when one of these modifiers is active:
- run2_miniAOD_devel
- run2_nanoAOD_102Xv1

the backport will depend on #25444

File: RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py [hunk 4]
```diff
@@ -29,4 +29,9 @@
                                                        872423215,872416066,872435036,872439336,
                                                        872420273,872436907,872420147,872439731,
                                                        872436657,872420397,872439732,872439339,
-                                                       872439603,872422436])
+                                                       872439603,872422436,872439861,872437051,
+                                                       872437052,872420649,872422436,872421950,
```
Comment (slava77) [inline RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py]: 872422436 is showing up twice.
Please check that this is just a redundancy and not a typo missing some other detid.

Comment (lathomas) [reply RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py]: @slava77  I just checked and I confirm that this is a redundancy, not a typo. This was the first "new" crystal spotted in 2018. We then converged on a more exhaustive list and for some reasons I copied this crystal twice. Sorry for that.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25436	550b24f
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences [some may be expected in the Flag_ecalBadCalibFilter decision, but the number of tested events is not that large to certainly expect a visible change]

**Pull request merged**